### DISCUSSION TITLE
sort | uniq -c equivalent

### DISCUSSION THREAD

**Discussion by alperyilmaz** [dathere/qsv#1677] [Q&A] [answered]
**Is your feature request related to a problem? Please describe.**
I want to count rows, just like `sort | uniq -c` does.. whatever in the row, maybe single value, or multiple values. Neither `qsv count` nor `qsv frequency` does this.

**Describe the solution you'd like**
Maybe `qsv count` might have `--rows` option. 

**Describe alternatives you've considered**
`sqlp` might be used for SQL based approach but I was looking for a practical `sort | uniq -c` alternative 

**Additional context**
One of the `qsv` functions should achieve the same output
```
$ echo -e "a,1\nb,2\na,1" | sort | uniq -c
      2 a,1
      1 b,2
```

**Comment by jqnatividad** [comment, marked as answer]
Hi @alperyilmaz ,

You can achieve this with `qsv frequency` by specifying `--no-headers` and a `--delimiter` that doesn't occur in the file:

```bash
$ echo -e "a,1\nb,2\na,1" | qsv frequency --no-headers --delimiter '|'
field,value,count
1,"a,1",2
1,"b,2",1

# to get exactly the output you want
$ echo -e "a,1\nb,2\na,1" | qsv frequency --no-headers --delimiter '|' \
  | qsv select count,value | qsv behead \
  | qsv fmt --quote-never --out-delimiter ' '
2 a,1
1 b,2
```

**Comment by ondohotola** [comment]
RTFM :-)-O

**Reply by alperyilmaz** [reply]
come on! I was just asking for a practical solution.. I tried frequency and it needed clean up. I usually use awk, and it's not practical. I was hoping for some practical counting with one or two commands.. I read the FM..

**Reply by 13minutes-yt** [reply]
try this in awk (or mawk for faster executions)
`echo -e "a,1\nb,2\na,1" |awk -F, '{a[$1","$2]++} END {for (i in a) print a[i],i }'`

**Reply by aborruso** [reply]
Using awk it's great in this way

```
echo -e "a,1\nb,2\na,1" | awk '!seen[$0]++'
```

**Comment by jqnatividad** [comment]
Moved it to Discussions as its the closest thing we have to a "Fine Manual" 😉

**Comment by aborruso** [comment]
Hi @alperyilmaz one tool that I use and find convenient is [Miller](https://miller.readthedocs.io/en/6.12.0/installing-miller/).

My sample input is this

```
a,1
b,2
a,1
```

If I run

```
mlr --nidx --ifs '|' count-distinct -f 1 input.txt
```

I get

```
a,1 2
b,2 1
```

I use "|" as separator, so long as it is not in the csv file, to have only one field (one field that contains the entire row).


I know, I am off-topic, but I am pleased to share a related experience.

**Comment by ggrothendieck** [comment]
Not as simple as you wanted, certainly what  you propose is simpler and I know you specifically said you did not want an SQL solution but nevertheless with SQL it is not that difficult either and it is only one line of code.  Suppose file1.csv contains:
```
Name,Value
a,1
b,2
a,1
```
Then
```
qsv sqlp file1.csv "select Name, Value, count(*) as n from file1 group by Name, Value order by Name, Value"
```
giving
```
Name,Value,n
a,1,2
b,2,1
(2, 3)
```

**Comment by ggrothendieck** [comment]
One other comment.  There is an huniq that supports huniq -c and is written in rust which would make it easier to incorporate into qsv or alternately it could be used independently of qsv.  

https://crates.io/crates/huniq
https://github.com/koraa/huniq

Example run
```
>qsv behead file1.csv | huniq -c
2 a,1
1 b,2
```
